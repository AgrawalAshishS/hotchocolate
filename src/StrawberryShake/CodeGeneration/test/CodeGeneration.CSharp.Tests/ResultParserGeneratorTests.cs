using System;
using System.Text;
using System.Threading.Tasks;
using Snapshooter.Xunit;
using Xunit;

namespace StrawberryShake.CodeGeneration.CSharp
{
    public class ResultParserGeneratorTests
    {
        [Fact]
        public async Task Generate_Nullable_List_Nullable_Element()
        {
            // arrange
            var sb = new StringBuilder();
            var writer = new CodeWriter(sb);

            var generator = new ResultParserGenerator();

            var descriptor = new ResultParserDescriptor(
                "GetHumanResultParser",
                "Demo",
                "IFoo",
                new[] {
                    new ResultParserMethodDescriptor(
                        "ParseFooBar",
                        "IBar",
                        new [] {
                            new ResultTypeDescriptor("Abc", true, true, true),
                            new ResultTypeDescriptor("Def", true, false, true)
                        },
                        false,
                        new [] {
                            new ResultFieldDescriptor("FieldA", "ParseThisAndThat")
                        }
                    ) },
                Array.Empty<ResultParserDeserializerMethod>(),
                Array.Empty<ValueSerializerDescriptor>());

            // act
            await generator.WriteAsync(writer, descriptor);

            // assert
            sb.ToString().MatchSnapshot();
        }

        [Fact]
        public async Task Generate_NonNull_List_Nullable_Element()
        {
            // arrange
            var sb = new StringBuilder();
            var writer = new CodeWriter(sb);

            var generator = new ResultParserGenerator();

            var descriptor = new ResultParserDescriptor(
                "GetHumanResultParser",
                "Demo",
                "IFoo",
                new[] {
                    new ResultParserMethodDescriptor(
                        "ParseFooBar",
                        "IBar",
                        new [] {
                            new ResultTypeDescriptor("Abc", false, true, true),
                            new ResultTypeDescriptor("Def", true, false, true)
                        },
                        false,
                        new [] {
                            new ResultFieldDescriptor("FieldA", "ParseThisAndThat")
                        }
                    ) },
                Array.Empty<ResultParserDeserializerMethod>(),
                Array.Empty<ValueSerializerDescriptor>());

            // act
            await generator.WriteAsync(writer, descriptor);

            // assert
            sb.ToString().MatchSnapshot();
        }

        [Fact]
        public async Task Generate_NonNull_List_NonNull_Element()
        {
            // arrange
            var sb = new StringBuilder();
            var writer = new CodeWriter(sb);

            var generator = new ResultParserGenerator();

            var descriptor = new ResultParserDescriptor(
                "GetHumanResultParser",
                "Demo",
                "IFoo",
                new[] {
                    new ResultParserMethodDescriptor(
                        "ParseFooBar",
                        "IBar",
                        new [] {
                            new ResultTypeDescriptor("Abc", false, true, true),
                            new ResultTypeDescriptor("Def", false, false, true)
                        },
                        false,
                        new [] {
                            new ResultFieldDescriptor("FieldA", "ParseThisAndThat")
                        }
                    ) },
                Array.Empty<ResultParserDeserializerMethod>(),
                Array.Empty<ValueSerializerDescriptor>());

            // act
            await generator.WriteAsync(writer, descriptor);

            // assert
            sb.ToString().MatchSnapshot();
        }

        [Fact]
        public async Task Generate_Nullable_List_Nullable_NestedList_Nullable_Element()
        {
            // arrange
            var sb = new StringBuilder();
            var writer = new CodeWriter(sb);

            var generator = new ResultParserGenerator();

            var descriptor = new ResultParserDescriptor(
                "GetHumanResultParser",
                "Demo",
                "IFoo",
                new[] {
                    new ResultParserMethodDescriptor(
                        "ParseFooBar",
                        "IBar",
                        new [] {
                            new ResultTypeDescriptor("Abc", true, true, true),
                            new ResultTypeDescriptor("Def", true, true, true),
                            new ResultTypeDescriptor("Ghi", true, false, true)
                        },
                        false,
                        new [] {
                            new ResultFieldDescriptor("FieldA", "ParseThisAndThat")
                        }
                    ) },
                Array.Empty<ResultParserDeserializerMethod>(),
                Array.Empty<ValueSerializerDescriptor>());

            // act
            await generator.WriteAsync(writer, descriptor);

            // assert
            sb.ToString().MatchSnapshot();
        }

        [Fact]
        public async Task Generate_NonNull_List_Nullable_NestedList_Nullable_Element()
        {
            // arrange
            var sb = new StringBuilder();
            var writer = new CodeWriter(sb);

            var generator = new ResultParserGenerator();

            var descriptor = new ResultParserDescriptor(
                "GetHumanResultParser",
                "Demo",
                "IFoo",
                new[] {
                    new ResultParserMethodDescriptor(
                        "ParseFooBar",
                        "IBar",
                        new [] {
                            new ResultTypeDescriptor("Abc", false, true, true),
                            new ResultTypeDescriptor("Def", true, true, true),
                            new ResultTypeDescriptor("Ghi", true, false, true)
                        },
                        false,
                        new [] {
                            new ResultFieldDescriptor("FieldA", "ParseThisAndThat")
                        }
                    ) },
                Array.Empty<ResultParserDeserializerMethod>(),
                Array.Empty<ValueSerializerDescriptor>());

            // act
            await generator.WriteAsync(writer, descriptor);

            // assert
            sb.ToString().MatchSnapshot();
        }

        [Fact]
        public async Task Generate_NonNull_List_NonNull_NestedList_Nullable_Element()
        {
            // arrange
            var sb = new StringBuilder();
            var writer = new CodeWriter(sb);

            var generator = new ResultParserGenerator();

            var descriptor = new ResultParserDescriptor(
                "GetHumanResultParser",
                "Demo",
                "IFoo",
                new[] {
                    new ResultParserMethodDescriptor(
                        "ParseFooBar",
                        "IBar",
                        new [] {
                            new ResultTypeDescriptor("Abc", false, true, true),
                            new ResultTypeDescriptor("Def", false, true, true),
                            new ResultTypeDescriptor("Ghi", true, false, true)
                        },
                        false,
                        new [] {
                            new ResultFieldDescriptor("FieldA", "ParseThisAndThat")
                        }
                    ) },
                Array.Empty<ResultParserDeserializerMethod>(),
                Array.Empty<ValueSerializerDescriptor>());

            // act
            await generator.WriteAsync(writer, descriptor);

            // assert
            sb.ToString().MatchSnapshot();
        }

        [Fact]
        public async Task Generate_NonNull_List_NonNull_NestedList_NonNull_Element()
        {
            // arrange
            var sb = new StringBuilder();
            var writer = new CodeWriter(sb);

            var generator = new ResultParserGenerator();

            var descriptor = new ResultParserDescriptor(
                "GetHumanResultParser",
                "Demo",
                "IFoo",
                new[] {
                    new ResultParserMethodDescriptor(
                        "ParseFooBar",
                        "IBar",
                        new [] {
                            new ResultTypeDescriptor("Abc", false, true, true),
                            new ResultTypeDescriptor("Def", false, true, true),
                            new ResultTypeDescriptor("Ghi", false, false, true)
                        },
                        false,
                        new [] {
                            new ResultFieldDescriptor("FieldA", "ParseThisAndThat")
                        }
                    ) },
                Array.Empty<ResultParserDeserializerMethod>(),
                Array.Empty<ValueSerializerDescriptor>());

            // act
            await generator.WriteAsync(writer, descriptor);

            // assert
            sb.ToString().MatchSnapshot();
        }

        [Fact]
        public async Task Generate_Nullable_Object()
        {
            // arrange
            var sb = new StringBuilder();
            var writer = new CodeWriter(sb);

            var generator = new ResultParserGenerator();

            var descriptor = new ResultParserDescriptor(
                "GetHumanResultParser",
                "Demo",
                "IFoo",
                new[] {
                    new ResultParserMethodDescriptor(
                        "ParseFooBar",
                        "IBar",
                        new [] {
                            new ResultTypeDescriptor("Ghi", true, false, true)
                        },
                        false,
                        new [] {
                            new ResultFieldDescriptor("FieldA", "ParseThisAndThat")
                        }
                    ) },
                Array.Empty<ResultParserDeserializerMethod>(),
                Array.Empty<ValueSerializerDescriptor>());

            // act
            await generator.WriteAsync(writer, descriptor);

            // assert
            sb.ToString().MatchSnapshot();
        }

        [Fact]
        public async Task Generate_NonNull_Object()
        {
            // arrange
            var sb = new StringBuilder();
            var writer = new CodeWriter(sb);

            var generator = new ResultParserGenerator();

            var descriptor = new ResultParserDescriptor(
                "GetHumanResultParser",
                "Demo",
                "IFoo",
                new[] {
                    new ResultParserMethodDescriptor(
                        "ParseFooBar",
                        "IBar",
                        new [] {
                            new ResultTypeDescriptor("Ghi", false, false, true)
                        },
                        false,
                        new [] {
                            new ResultFieldDescriptor("FieldA", "ParseThisAndThat")
                        }
                    ) },
                Array.Empty<ResultParserDeserializerMethod>(),
                Array.Empty<ValueSerializerDescriptor>());

            // act
            await generator.WriteAsync(writer, descriptor);

            // assert
            sb.ToString().MatchSnapshot();
        }

        [Fact]
        public async Task Generate_Deserialize_Nullable_String_Deserializer()
        {
            // arrange
            var sb = new StringBuilder();
            var writer = new CodeWriter(sb);

            var generator = new ResultParserGenerator();

            var descriptor = new ResultParserDescriptor(
                "GetHumanResultParser",
                "Demo",
                "IFoo",
                Array.Empty<ResultParserMethodDescriptor>(),
                new[] {
                    new ResultParserDeserializerMethod(
                        "DeserializeNullableString",
                        "string",
                        "string",
                        new [] {
                            new ResultTypeDescriptor("string", true, false, true)
                        },
                        new ValueSerializerDescriptor("String", "_stringSerializer"))
                },
                new[] {
                    new ValueSerializerDescriptor("String", "_stringSerializer")
                });

            // act
            await generator.WriteAsync(writer, descriptor);

            // assert
            sb.ToString().MatchSnapshot();
        }

        [Fact]
        public void CanHandle()
        {
            // arrange
            var generator = new ResultParserGenerator();

            var descriptor = new ResultParserDescriptor(
                "GetHumanResultParser",
                "Demo",
                "IFoo",
                new[] {
                    new ResultParserMethodDescriptor(
                        "ParseFooBar",
                        "IBar",
                        new [] {
                            new ResultTypeDescriptor("Ghi", false, false, true)
                        },
                        false,
                        new [] {
                            new ResultFieldDescriptor("FieldA", "ParseThisAndThat")
                        }
                    ) },
                Array.Empty<ResultParserDeserializerMethod>(),
                Array.Empty<ValueSerializerDescriptor>());

            // act
            var canHandle = generator.CanHandle(descriptor);

            // assert
            Assert.True(canHandle);
        }
    }
}
