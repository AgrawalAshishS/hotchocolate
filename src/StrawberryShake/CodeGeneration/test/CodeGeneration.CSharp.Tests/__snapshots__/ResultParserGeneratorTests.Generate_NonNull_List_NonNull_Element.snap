namespace Demo
{
    [global::System.CodeDom.Compiler.GeneratedCode("StrawberryShake", "11.0.0")]
    public partial GetHumanResultParser
        : global::StrawberryShake.Http.JsonResultParserBase<IFoo>
    {
        public GetHumanResultParser(global::StrawberryShake.IValueSerializerCollection serializerResolver)
        {
            if (serializerResolver is null)
            {
                throw new global::System.ArgumentNullException(nameof(serializerResolver));
            }

        }

        private override IBar? ParseFooBar(global::System.Text.Json.JsonElement parent)
        {
            if (!parent.TryGetProperty(field, out global::System.Text.Json.JsonElement obj)
                || obj.ValueKind == global::System.Text.Json.JsonValueKind.Null)
            {
                throw new global::System.InvalidOperationException(field);
            }

            int count = obj.GetArrayLength();
            var result = new Abc[count];

            for (int i = 0; i < count; i++)
            {
                global::System.Text.Json.JsonElement element = obj[i];

                result[i] = new Def
                (
                    ParseThisAndThat(element, "FieldA")
                );
            }

            return result;
        }
    }
}
